File: 0373-find-k-pairs-with-smallest-sums/0373-find-k-pairs-with-smallest-sums.java
class Solution {
    public List<List<Integer>> kSmallestPairs(int[] nums1, int[] nums2, int k) {
        PriorityQueue<int[]> minheap=new PriorityQueue<>((a,b)->a[0]-b[0]);
        for(int i=0;i<nums2.length;i++){
            int sum=nums1[0]+nums2[i];
            minheap.offer(new int[]{sum,0,i});
        }
        List<List<Integer>> res=new ArrayList<>();
        while(k-->0){
            int[] x=new int[3];
            x=minheap.poll();
            List<Integer> ls=new ArrayList<>();
            ls.addAll(Arrays.asList(nums1[x[1]], nums2[x[2]]));
            res.add(ls);
            if(x[1]+1<nums1.length){
                int sum2=nums1[x[1]+1]+nums2[x[2]];
                minheap.offer(new int[]{sum2,x[1]+1,x[2]});
            }
        }
        return res;
    }
}

File: 0373-find-k-pairs-with-smallest-sums/0373-find-k-pairs-with-smallest-sums.py
class Solution(object):
    def kSmallestPairs(self, nums1, nums2, k):
        def helper(nums1,nums2,k):
            heap=[]
            for i in range(len(nums2)):
                sum=nums1[0] + nums2[i]
                heapq.heappush(heap,(sum,0,i))

            res=[]
            while k>0:
                sum,i,j=heapq.heappop(heap)
                res.append([nums1[i],nums2[j]])
                if i+1<len(nums1):
                    heapq.heappush(heap,(nums1[i+1]+nums2[j],i+1,j))
                k-=1

            return res
        return helper(nums1,nums2,k)
        